streams_map.go

```diff
@@ -1,234 +1,175 @@
 package quic
 
 import (
-	"errors"
 	"fmt"
-	"sync"
 
+	"github.com/lucas-clemente/quic-go/internal/flowcontrol"
 	"github.com/lucas-clemente/quic-go/internal/handshake"
 	"github.com/lucas-clemente/quic-go/internal/protocol"
-	"github.com/lucas-clemente/quic-go/internal/wire"
-	"github.com/lucas-clemente/quic-go/qerr"
 )
 
-type streamsMap struct {
-	mutex sync.RWMutex
-
-	perspective protocol.Perspective
+type streamType int
 
-	streams map[protocol.StreamID]streamI
+const (
+	streamTypeOutgoingBidi streamType = iota
+	streamTypeIncomingBidi
+	streamTypeOutgoingUni
+	streamTypeIncomingUni
+)
 
-	nextStreamToOpen          protocol.StreamID // StreamID of the next Stream that will be returned by OpenStream()
-	highestStreamOpenedByPeer protocol.StreamID
-	nextStreamOrErrCond       sync.Cond
-	openStreamOrErrCond       sync.Cond
+type streamsMap struct {
+	perspective protocol.Perspective
 
-	closeErr           error
-	nextStreamToAccept protocol.StreamID
+	sender            streamSender
+	newFlowController func(protocol.StreamID) flowcontrol.StreamFlowController
 
-	newStream newStreamLambda
+	outgoingBidiStreams *outgoingBidiStreamsMap
+	outgoingUniStreams  *outgoingUniStreamsMap
+	incomingBidiStreams *incomingBidiStreamsMap
+	incomingUniStreams  *incomingUniStreamsMap
```

Comment:
Some of the functions here could be simpler if you made this a map[streamType]commonInterface – just a thought if that didn't occur to you.

Reply:
There no overlapping interface between incoming and outgoing stream maps.

Reply:
DeleteStream is common. Not important though.